packages/api-server/src/price-oracle.ts

```diff
@@ -0,0 +1,143 @@
+const PRICE_DIFF = 0.05; // 5%
+const PRICE_UPDATE_WINDOW = 5 * 60000; // 5 minutes
+const CKB_PRICE_CACHE_KEY = "priceOracle:ckbUsd";
+
+export class CKBPriceOracle extends BaseWorker {
+  private cacheStore: Store;
+  private readonly: boolean;
+
+  constructor({
+    readonly = false,
+    pollTimeInterval = POLL_TIME_INTERVAL,
+    livenessCheckInterval = LIVENESS_CHECK_INTERVAL,
+    expiredTime = CACHE_EXPIRED_TIME,
+  } = {}) {
+    super({ pollTimeInterval, livenessCheckInterval });
+    this.cacheStore = new Store(true, expiredTime);
+    this.readonly = readonly;
+  }
+
+  startForever(): Promise<void> {
+    if (this.readonly) {
+      throw new Error("readonly oracle cannot start forever working");
+    }
+    return super.startForever();
+  }
+
+  start(): Promise<void> {
+    if (this.readonly) {
+      throw new Error("readonly oracle cannot start working");
+    }
+    return super.start();
+  }
+
+  protected async executePoll(): Promise<number> {
+    const [newPrice, price, ttl] = await Promise.all([
+      this.pollPrice(),
+      this.cacheStore.get(CKB_PRICE_CACHE_KEY),
+      this.cacheStore.ttl(CKB_PRICE_CACHE_KEY),
+    ]);
+
+    if (
+      price == null ||
+      ttl <= CACHE_EXPIRED_TIME - PRICE_UPDATE_WINDOW ||
+      calcDiff(+newPrice, +price) > PRICE_DIFF
+    ) {
+      await this.cacheStore.insert(CKB_PRICE_CACHE_KEY, newPrice);
+      return this.pollTimeInterval;
+    }
+
+    return this.pollTimeInterval;
+  }
+
+  async price(): Promise<string> {
+    const price = await this.cacheStore.get(CKB_PRICE_CACHE_KEY);
+    if (price == null) {
+      return await this.pollPrice();
+    }
+    return price;
+  }
+
+  async minGasPrice(): Promise<bigint> {
+    const ckbPrice = await this.price();
+    return Price.from(ckbPrice).toMinGasPrice();
+  }
+
+  async minFeeRate(): Promise<bigint> {
+    const ckbPrice = await this.price();
+    return Price.from(ckbPrice).toMinFeeRate();
+  }
+
+  private async sendRequest(url: string) {
+    let response = await fetch(url);
+    if (response.status === 200) {
+      let data = await response.text();
+      return data;
+    } else {
+      throw new Error("fetch failed, error code " + response.status);
+    }
+  }
+
+  private async pollPrice(): Promise<string> {
+    // rate-limit: 50 requests 1 minute
+    const coingecko = async () => {
+      const tokenId = "nervos-network";
+      const url = `https://api.coingecko.com/api/v3/simple/price?ids=${tokenId}&vs_currencies=usd`;
+      const res = await this.sendRequest(url);
+      const resObj = JSON.parse(res);
+      if (!("usd" in resObj[tokenId])) {
+        throw new Error(`request to ${url} error, result: ${res}`);
+      }
+      return new Decimal(resObj[tokenId].usd).toString();
+    };
+
+    // rate-limit: 160000 requests 1 day, 50 requests 10 seconds
+    const binance = async () => {
+      const symbol = "CKBUSDT";
+      const url = `https://api.binance.com/api/v3/trades?symbol=${symbol}&limit=1`;
```

Comment:
I'm having trouble connecting to api.binance.com from my computer. Then I received a lot of noisy error logs from my terminal.
It's worth better handling connection errors.

Reply:
It should be part of healthcheck and be covered in `poly_getHealthStatus`.

Reply:
we check connection [status code](https://github.com/nervosnetwork/godwoken-web3/pull/514/files#diff-b2e875bea75f54d5de088e6db11984598f7942fdb526a1c2327f4ab6bbe66fceR85) in `sendRequest`and data format [in here](https://github.com/nervosnetwork/godwoken-web3/pull/514/files#diff-b2e875bea75f54d5de088e6db11984598f7942fdb526a1c2327f4ab6bbe66fceR115), throw error if anything happen.  Any suggestions for more error handling?

Reply:
It's a difficult problem...

What about dropping the failed provider(it would not affect our service)? We will/should receive alerts from a platform that calls `poly_getHealthStatus`

Reply:
1. it will affect service
2. the price APIs have rate limit, so exposing the call to connect to them in public rpc `poly_getHealthStatus` might not be ideal.
3. one simple idea: the price oracle should keep updating the `ckb price` in redis, otherwise it will be expired. We can check if such redis key value is available in `poly_getHealthStatus`, if not, the service is not health and something wrong with the price oracle.

Reply:
> 3. one simple idea: the price oracle should keep updating the `ckb price` in redis, otherwise it will be expired. We can check if such redis key value is available in `poly_getHealthStatus`, if not, the service is not health and something wrong with the price oracle.

Cool. I like this idea!